core/src/main/java/org/elasticsearch/search/SearchModule.java

```diff
@@ -364,16 +377,28 @@ protected void configureFetchSubPhase() {
 
     public IndicesQueriesRegistry buildQueryParserRegistry() {
         Map<String, QueryParser<?>> queryParsersMap = new HashMap<>();
+
+        // TODO remove this when we retire registerQueryParser
         for (Supplier<QueryParser<?>> parserSupplier : queryParsers) {
-            QueryParser<? extends QueryBuilder> parser = parserSupplier.get();
+            QueryParser<? extends QueryBuilder<?>> parser = parserSupplier.get();
             for (String name: parser.names()) {
                 Object oldValue = queryParsersMap.putIfAbsent(name, parser);
                 if (oldValue != null) {
                     throw new IllegalArgumentException("Query parser [" + oldValue + "] already registered for name [" + name + "] while trying to register [" + parser + "]");
                 }
             }
-            @SuppressWarnings("unchecked") NamedWriteable<? extends QueryBuilder> qb = parser.getBuilderPrototype();
-            namedWriteableRegistry.registerPrototype(QueryBuilder.class, qb);
+            namedWriteableRegistry.registerPrototype(QueryBuilder.class, parser.getBuilderPrototype());
+        }
+
+        for (QueryRegistration<?> query : queries) {
```

Comment:
I wonder if later on we are able to get rid of QueryRegistration and register stuff straight-away rather than when calling buildQueryParserRegistry. we will see later though!